### PR TITLE
Update xml classes of Serial, Console and Channel

### PR DESCRIPTION
**Pull request by chloerh**
To support source as nested xml.

Signed-off-by: Haijiao Zhao <haizhao@redhat.com>

### CODE REVIEW

File: virttest/libvirt_xml/devices/character.py [hunk 1]
```diff
@@ -74,3 +62,83 @@ def update_target(self, index, **attributes):
...
+    @staticmethod
+    def marshal_from_sources(item, index, libvirtxml):
+        """
+        Convert an xml object to source tag and xml element.
+        """
+        if isinstance(item, CharacterBase.Source):
+            print('1')
+            return 'source', item
+        elif isinstance(item, dict):
+            print('2')
```
Comment (Yingshun) [inline virttest/libvirt_xml/devices/character.py:75]: Should we remove 'print()'?

**Review by Yingshun — COMMENTED** [review]
Others LGTM

File: virttest/libvirt_xml/devices/character.py (hunk 1 above)
Comment (chloerh) [reply virttest/libvirt_xml/devices/character.py]: Updated.

**Review by Yingshun — APPROVED** [review]

**Review by dzhengfy — COMMENTED** [review]
LGTM. But the changes seem complicated and impact scope is big. Could you give a test information like python interactive session test output and sample xmls for better understanding?

**Comment by chloerh** [comment]
Depends on 
- https://github.com/avocado-framework/avocado-vt/pull/3526

**Comment by chloerh** [comment]
```
>>>from virttest.libvirt_xml.devices import channel, serial, console
>>>seri = serial.Serial()
>>>seri.setup_attrs(**{'target_port': '1', 'targets': [{'port': '1'}], 'type_name': 'udp', 'sources': [{'attrs': {'host': '0.0.0.0', 'mode': 'bind', 'service': '2445'}}, {'attrs': {'host': '0.0.0.0', 'mode': 'connect', 'service': '2445'}}]})
>>>
>>>print(seri)
<?xml version='1.0' encoding='UTF-8'?>
<serial type="udp"><target port="1" /><source host="0.0.0.0" mode="bind" service="2445" /><source host="0.0.0.0" mode="connect" service="2445" /></serial>
```

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**